libobs/audio-monitoring/nix/pulseaudio-output.c

```diff
@@ -0,0 +1,487 @@
+#include "obs-internal.h"
+#include "pulseaudio-wrapper.h"
+
+#define PULSE_DATA(voidptr) struct audio_monitor *data = voidptr;
+#define blog(level, msg, ...) blog(level, "pulse-am: " msg, ##__VA_ARGS__)
+
+struct audio_monitor {
+	obs_source_t *source;
+	pa_stream *stream;
+	char *device;
+
+	enum speaker_layout speakers;
+	pa_sample_format_t format;
+	uint_fast32_t samples_per_sec;
+	uint_fast32_t bytes_per_frame;
+	uint_fast8_t channels;
+
+	uint_fast32_t packets;
+	uint_fast64_t frames;
+
+	struct circlebuf new_data;
+	audio_resampler_t *resampler;
+	size_t buffer_size;
+	size_t bytesRemaining;
+
+	size_t bytes_per_channel;
+	bool ignore : 1;
+	pthread_mutex_t playback_mutex;
+};
+
+static enum speaker_layout pulseaudio_channels_to_obs_speakers(
+	uint_fast32_t channels)
+{
+	if ((channels >= 1 && channels <= 6) || channels == 8)
+		return (enum speaker_layout) channels;
+	return SPEAKERS_UNKNOWN;
+}
+
+static enum audio_format pulseaudio_to_obs_audio_format(
+	pa_sample_format_t format)
+{
+	switch (format) {
+	case PA_SAMPLE_U8:
+		return AUDIO_FORMAT_U8BIT;
+	case PA_SAMPLE_S16LE:
+		return AUDIO_FORMAT_16BIT;
+	case PA_SAMPLE_S32LE:
+		return AUDIO_FORMAT_32BIT;
+	case PA_SAMPLE_FLOAT32LE:
+		return AUDIO_FORMAT_FLOAT;
+	default:
+		return AUDIO_FORMAT_UNKNOWN;
+	}
+}
+
+static void process_byte(void *p, size_t frames, size_t channels, float vol)
+{
+	register char *cur = (char *) p;
+	register char *end = cur + frames * channels;
+
+	while (cur < end)
+		*(cur++) *= vol;
+}
+
+static void process_short(void *p, size_t frames, size_t channels, float vol)
+{
+	register short *cur = (short *) p;
+	register short *end = cur + frames * channels;
+
+	while (cur < end)
+		*(cur++) *= vol;
+}
+
+static void process_float(void *p, size_t frames, size_t channels, float vol)
+{
+	register float *cur = (float *) p;
+	register float *end = cur + frames * channels;
+
+	while (cur < end)
+		*(cur++) *= vol;
+}
+
+void process_volume(const struct audio_monitor *monitor, float vol,
+                    uint8_t *const *resample_data, uint32_t resample_frames)
+{
+	switch (monitor->bytes_per_channel) {
+	case 1:
+		process_byte(resample_data[0], resample_frames,
+		             monitor->channels, vol);
+		break;
+	case 2:
+		process_short(resample_data[0], resample_frames,
+		              monitor->channels, vol);
+		break;
+	default:
+		process_float(resample_data[0], resample_frames,
+		              monitor->channels, vol);
+		break;
+	}
+}
+
+static void do_stream_write(void *param)
+{
+	PULSE_DATA(param);
+	uint8_t *buffer = NULL;
+
+	while (data->new_data.size >= data->buffer_size &&
+	       data->bytesRemaining > 0) {
```

Comment:
Style:  Indent with one tab for continuing a control structure or loop statement (if, while, for, etc.).